Can't boot Keycloak server with Java KeyStore p12 file running on Windows 10/2019

### Before reporting an issue

- [X] I have searched existing issues
- [ ] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

dist/quarkus

### Describe the bug

Running Keycloak on Windows 10/2019. I whant to secure MSSQL database password in keyclock.conf file using an Java Keystore p12 file. When booting Keycloak server from PowerShell with command:  **.\kc.bat start-dev --config-keystore=/keystore.p12 --config-keystore-password=storepass --config-keystore-type=PKCS12** I'm getting the following error message: **_Error: java.lang.IllegalArgumentException: SRCFG00034: URI Syntax invalid C:\keystore.p12_**

Stack trace:

PS C:\KeyCloak\keycloak-22.0.1\bin> .\kc.bat start-dev --config-keystore=/keystore.p12 --config-keystore-password=storepass --config-keystore-type=PKCS12
The DelayedHandler was closed before any children handlers were configured. Messages will be written to stderr.
2023-08-02 09:59:18,613 DEBUG [org.jboss.logging] (main) Logging Provider: org.jboss.logging.JBossLogManagerProvider

Exception in thread "main" java.lang.reflect.InvocationTargetException
at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:119)
at java.base/java.lang.reflect.Method.invoke(Method.java:578)
at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
Caused by: java.lang.IllegalArgumentException: SRCFG00034: URI Syntax invalid C:\keystore.p12
at io.smallrye.config.AbstractLocationConfigSourceLoader$URIConverter.convert(AbstractLocationConfigSourceLoader.java:311)
at io.smallrye.config.AbstractLocationConfigSourceLoader$URIConverter.convert(AbstractLocationConfigSourceLoader.java:303)
at io.smallrye.config.AbstractLocationConfigSourceLoader.loadConfigSources(AbstractLocationConfigSourceLoader.java:99)
at io.smallrye.config.AbstractLocationConfigSourceLoader.loadConfigSources(AbstractLocationConfigSourceLoader.java:89)
at io.smallrye.config.AbstractLocationConfigSourceLoader.loadConfigSources(AbstractLocationConfigSourceLoader.java:81)
at io.smallrye.config.source.keystore.KeyStoreConfigSourceFactory$1.getConfigSources(KeyStoreConfigSourceFactory.java:49)
at io.smallrye.config.ConfigurableConfigSource.getConfigSources(ConfigurableConfigSource.java:50)
at io.smallrye.config.ConfigurableConfigSource.getConfigSources(ConfigurableConfigSource.java:52)
at io.smallrye.config.SmallRyeConfig$ConfigSources.mapLateSources(SmallRyeConfig.java:669)
at io.smallrye.config.SmallRyeConfig$ConfigSources.(SmallRyeConfig.java:553)
at io.smallrye.config.SmallRyeConfig.(SmallRyeConfig.java:69)
at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:629)
at io.quarkus.runtime.configuration.QuarkusConfigFactory.getConfigFor(QuarkusConfigFactory.java:29)
at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:76)
at io.smallrye.config.SmallRyeConfigProviderResolver.getConfig(SmallRyeConfigProviderResolver.java:64)
at org.keycloak.quarkus.runtime.configuration.Configuration.getConfig(Configuration.java:51)
at org.keycloak.quarkus.runtime.cli.Picocli.hasConfigChanges(Picocli.java:239)
at org.keycloak.quarkus.runtime.cli.Picocli.requiresReAugmentation(Picocli.java:141)
at org.keycloak.quarkus.runtime.cli.Picocli.runReAugmentationIfNeeded(Picocli.java:126)
at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:95)
at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:88)
at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
... 3 more
Caused by: java.net.URISyntaxException: Illegal character in opaque part at index 2: C:\keystore.p12
at java.base/java.net.URI$Parser.fail(URI.java:2970)
at java.base/java.net.URI$Parser.checkChars(URI.java:3141)
at java.base/java.net.URI$Parser.parse(URI.java:3177)
at java.base/java.net.URI.(URI.java:620)
at io.smallrye.config.AbstractLocationConfigSourceLoader$URIConverter.convert(AbstractLocationConfigSourceLoader.java:309)
... 24 more

I have tried to format the URI in --config-keystore=/keystore.p12 but no luck :( any suggestions?

### Version

22.0.1

### Expected behavior

KeyCloack server must boot and connect to MSSQL database server using an secured Java KeyStore file fro database password. Have used the following guide: https://www.keycloak.org/server/configuration#_setting_sensitive_options_using_a_java_keystore_file

Customer don't accept password written in clear text, within config files (in this case it's keycloak.conf).

 

### Actual behavior

When I boot Keycloak server it won't boot cause of the followoing error:

**_Error: java.lang.IllegalArgumentException: SRCFG00034: URI Syntax invalid C:\keystore.p12_**

I have tried to change the URI in the startup command parameter: **--config-keystore=/keystore.p12** in command: **.\kc.bat start-dev --config-keystore=/keystore.p12 --config-keystore-password=storepass --config-keystore-type=PKCS12** but with no luck :(

### How to Reproduce?

Follow guide here: https://www.keycloak.org/server/configuration#_setting_sensitive_options_using_a_java_keystore_file:
 
Running commands in PowerShell:

**keytool -importpass -alias kc.db-password -keystore keystore.p12 -storepass keystorepass -storetype PKCS12 -v**

and

 **.\kc.bat start-dev --config-keystore=/keystore.p12 --config-keystore-password=storepass --config-keystore-type=PKCS12**

On Windows environment. 

### Anything else?

Discussion; https://github.com/keycloak/keycloak/discussions/22156 #22156

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -126,7 +126,10 @@ jobs:
     name: Quarkus UT
     needs: build
     timeout-minutes: 15
-    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os: [ ubuntu-latest, windows-latest ]
+    runs-on: ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v4
 

--- a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/ConfigKeystorePropertyMappers.java
+++ b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/ConfigKeystorePropertyMappers.java
@@ -6,7 +6,6 @@ import org.keycloak.config.ConfigKeystoreOptions;
 
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.nio.file.Paths;
 import java.util.Optional;
 
 import static org.keycloak.quarkus.runtime.configuration.mappers.PropertyMapper.fromOption;
@@ -50,11 +49,12 @@ final class ConfigKeystorePropertyMappers {
             throw new IllegalArgumentException("config-keystore-password must be specified");
         }
 
-        Optional<String> realPath = Optional.of(String.valueOf(Paths.get(path.getValue()).toAbsolutePath().normalize()));
-        if (!Files.exists(Path.of(realPath.get()))) {
-            throw new IllegalArgumentException("config-keystore path does not exist: " + realPath.get());
+        final Path realPath = Path.of(path.getValue()).toAbsolutePath().normalize();
+        if (!Files.exists(realPath)) {
+            throw new IllegalArgumentException("config-keystore path does not exist: " + realPath);
         }
-        return realPath;
+
+        return Optional.of(realPath.toUri().toString());
     }
 
     private static Optional<String> validatePassword(Optional<String> option, ConfigSourceInterceptorContext context) {

--- a/quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/test/ConfigurationTest.java
+++ b/quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/test/ConfigurationTest.java
@@ -20,10 +20,11 @@ package org.keycloak.quarkus.runtime.configuration.test;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
+import static org.keycloak.quarkus.runtime.Environment.isWindows;
 import static org.keycloak.quarkus.runtime.configuration.ConfigArgsConfigSource.CLI_ARGS;
 
-import java.io.File;
 import java.lang.reflect.Field;
+import java.nio.file.Path;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Properties;
@@ -266,7 +267,14 @@ public class ConfigurationTest {
         System.setProperty(CLI_ARGS, "--db=dev-file");
         SmallRyeConfig config = createConfig();
         assertEquals(H2Dialect.class.getName(), config.getConfigValue("kc.db-dialect").getValue());
-        assertEquals("jdbc:h2:file:" + System.getProperty("user.home") + "/data/h2/keycloakdb;;AUTO_SERVER=TRUE;NON_KEYWORDS=VALUE", config.getConfigValue("quarkus.datasource.jdbc.url").getValue());
+
+        // JDBC location treated as file:// URI
+        final String userHomeUri = Path.of(System.getProperty("user.home"))
+                .toUri()
+                .toString()
+                .replaceFirst(isWindows() ? "file:///" : "file://", "");
+
+        assertEquals("jdbc:h2:file:" + userHomeUri + "data/h2/keycloakdb;;AUTO_SERVER=TRUE;NON_KEYWORDS=VALUE", config.getConfigValue("quarkus.datasource.jdbc.url").getValue());
 
         System.setProperty(CLI_ARGS, "--db=dev-mem");
         config = createConfig();

